### PR TITLE
Fixed field shadows being drawn over 3D models

### PR DESCRIPTION
**Pull request by CosmosXIII**
Fixed the field shadows being drawn over the 3d models. I also removed all the stencil buffer related API since those are not needed anymore.

### CODE REVIEW

File: src/gl/gl.cpp [hunk 1]
```diff
@@ -246,6 +246,12 @@ void gl_draw_indexed_primitive(uint32_t primitivetype, uint32_t vertextype, stru
 
 	newRenderer.setInterpolationQualifier(current_state.shademode ? RendererInterpolationQualifier::SMOOTH : RendererInterpolationQualifier::FLAT);
 
+	if (vertextype != TLVERTEX)
+	{
+		newRenderer.setD3DProjection(&current_state.d3dprojection_matrix);
+		newRenderer.setD3DViweport(&d3dviewport_matrix);
+	}
+
```
Comment (julianxhokaxhiu) [inline src/gl/gl.cpp:254]: May I ask why did you move this block? Is this really required?

File: src/gl.h [hunk 2]
```diff
@@ -93,7 +103,7 @@ void gl_save_state(struct driver_state *dest);
 void gl_load_state(struct driver_state *src);
 uint32_t gl_defer_draw(uint32_t primitivetype, uint32_t vertextype, struct nvertex* vertices, struct point3d* normals, uint32_t vertexcount, WORD* indices, uint32_t count, struct boundingbox* boundingbox, uint32_t clip, uint32_t mipmap);
 uint32_t gl_defer_sorted_draw(uint32_t primitivetype, uint32_t vertextype, struct nvertex *vertices, uint32_t vertexcount, WORD *indices, uint32_t count, uint32_t clip, uint32_t mipmap);
-void gl_draw_deferred(bool isDrawOpaqueOnly);
+void gl_draw_deferred(int draworder);
```
Comment (julianxhokaxhiu) [inline src/gl.h:106]: Use the Enum type directly if you expect to be always that one

File: src/gl/deferred.cpp [hunk 3]
```diff
@@ -291,7 +296,7 @@ uint32_t gl_defer_sorted_draw(uint32_t primitivetype, uint32_t vertextype, struc
 }
 
 // draw deferred models
-void gl_draw_deferred(bool isDrawOpaqueOnly)
+void gl_draw_deferred(int draworder)
```
Comment (julianxhokaxhiu) [inline src/gl/deferred.cpp:299]: Same as stated above

**Review by julianxhokaxhiu — CHANGES_REQUESTED** [review]
Overall looks good, there are minor things to clarify and to amend. Great job so far!

File: src/gl.h (hunk 2 above)
Comment (CosmosXIII) [reply src/gl.h]: Changet it to use the Enum as parameter.

File: src/gl/deferred.cpp (hunk 3 above)
Comment (CosmosXIII) [reply src/gl/deferred.cpp]: Changed it to use the Enum as parameter.

File: src/gl/gl.cpp (hunk 1 above)
Comment (CosmosXIII) [reply src/gl/gl.cpp:254]: I moved this code to guarantee that the projection and viewport matrices are set before drawing the field shadows. Until now I was drawing other 3d models such as the characters before the field shadows so these matrices were set before drawing the shadows. However, with the changes of this PR, the field shadows (walkmesh) are the first 3d object to be drawn in the frame. By moving this block of code before the draw call deferring I was able to guarantee that these matrices are set before any drawing. Without this change the matrices used when drawing the field shadows have a one frame delay.

As far as I know these matrices are common for all 3d objects so it should be enough to set once at the beginning of the frame. However, I did it this way because it was the method with the less amount of changes required. If this change is not desired I could work around it by adding a new method that sets the matrices for the first 3d object before any drawing. Let me know what you think and thanks for the review as always!

Comment (julianxhokaxhiu) [reply src/gl/gl.cpp:254]: To be honest the only part that concerns me is that this code is in common with FF8 which has a tons of tricks in place put here and there to guarantee perfect rendering. This part of the code is one of those and I have no idea if this code change might have an impact ( as you notice there's an if case before this block originally which might return and set basically nothing. Moving this block before that if, now might bring some implications i might not be aware.

If you can find a way to set those in your lighting code class I would appreciate. Thank you!

Comment (CosmosXIII) [reply src/gl/gl.cpp]: I implemented an alternative way to set the matrices using a new method that is only called when lighting is enabled.

Comment (julianxhokaxhiu) [reply src/gl/gl.cpp]: Amazing, thank you :)

**Review by julianxhokaxhiu — APPROVED** [review]
